distributed/s3.py

```diff
@@ -65,6 +66,66 @@ def get_list_of_summary_objects(bucket_name, prefix='', delimiter='',
     return [s for s in L if s.key[-1] != '/']
 
 
+def seek_delimiter(ob, offset, delimiter, seeklength=4096):
+    """
+    Find the bytes location of the next delimiter after the given offset
+
+    Parameters
+    ----------
+    ob : boto3 open s3 key
+    offset : int
+        bytes location to start
+    delimiter : bytes
+        pattern to seek
+    seeklength : int
+        If seeking delimited, the blocksize for finding delimiters.
+    """
+    if offset == 0:
+        return 0
+    data = b''
+    while True:
+        try:
+            piece = ob.get(Range="bytes=%i-%i" %
+                           (offset, offset + seeklength))['Body'].read()
+        except ClientError:
+            return ob.content_length
+        data += piece
+        try:
+            i = data.index(delimiter)
+            return offset + i + len(delimiter)
+        except ValueError:
+            offset += seeklength
+
+
+def read_block(buc, fn, offset, length, delimiter=None, anon=None):
+    """
+    Read one chunk of bytes from the given S3 key, optionally delimited.
+
+    Parameters
+    ----------
+    ob : boto3 open s3 key
+    offset : int
+        From which byte location to read
+    length : int
+        size of block to read
+    delimiter : bytes
+        If given, seek to next delimiter at the star and end of the block
+    """
+    s3 = get_s3(anon)
+    ob = s3.Object(buc, fn)
+    if delimiter is None:
+        try:
+            return ob.get(Range="bytes=%i-%i" %
+                          (offset, offset + length - 1))['Body'].read()
+        except ClientError:
+            return b''
+    start = seek_delimiter(ob, offset, delimiter)
+    end = seek_delimiter(ob, offset + length, delimiter)
+    if start == end:
+        return b''
+    return ob.get(Range="bytes=%i-%i" % (start, end - 1))['Body'].read()
```

Comment:
If each individual read is expensive then it might make sense to initially read most of the bytes straight away with a big buffer on either side.  E.g., if I ask for the byte range `1000000->2000000`, then perhaps we just read `900000->2100000` straightaway and only go seeking if we don't find a delimiter.